eliona/assets.go

```diff
@@ -85,3 +85,61 @@ func name(specification kentix.DeviceInfo) string {
 func description(specification kentix.DeviceInfo) string {
 	return fmt.Sprintf("%s (%s)", specification.Name, specification.Serial)
 }
+
+// CreateDoorlockAssetsIfNecessary create all assets for specification if not already exists
+func CreateDoorlockAssetsIfNecessary(config apiserver.Configuration, spec kentix.DoorLock) error {
+	for _, projectId := range conf.ProjIds(config) {
+		if err := createDoorlockAssetIfNecessary(config, projectId, nil, spec); err != nil {
+			return fmt.Errorf("creating assets for device %s: %v", spec.Serial, err)
+		}
+	}
+	return nil
+}
+
+// createDoorlockAssetIfNecessary create asset for specification if not already exists
+func createDoorlockAssetIfNecessary(config apiserver.Configuration, projectId string, parentAssetId *int32, spec kentix.DoorLock) error {
+	// Get known asset id from configuration
+	assetID, err := conf.GetAssetId(context.Background(), config, projectId, spec.Serial)
+	if err != nil {
+		return fmt.Errorf("finding asset ID: %v", err)
+	}
+	if assetID != nil {
+		return nil
+	}
+
+	// If no asset exists for project and configuration, create a new one
+	name := doorlockName(spec)
+	description := doorlockDescription(spec)
+
+	newId, err := asset.UpsertAsset(api.Asset{
+		ProjectId:               projectId,
+		GlobalAssetIdentifier:   spec.Serial,
+		Name:                    *api.NewNullableString(common.Ptr(name)),
+		AssetType:               kentix.DoorlockAssetType,
+		Description:             *api.NewNullableString(common.Ptr(description)),
+		ParentLocationalAssetId: *api.NewNullableInt32(parentAssetId),
```

Comment:
There are 2 types of relationship: Local and functional. I think, in this case the functional relationship is the better one -> parentFunctionalAssetId

Some other words to illustrate the difference: **local** = geographic positions e.g. a door  and a room. You can not change the place of the door in relation to the room. **functional** = logical dependency e.g. door and keycard to open it. The keycard's position is not fix, but it belongs to the door.

For Kentix: the doorlock should have the AccessManager as functional parant. In addition the local parent can be a Room asset.

Reply:
Aha, thanks for the explanation.